lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+import jwt
+from flask import request, jsonify, make_response
+from flask_restful import Resource, abort
+from flask_bcrypt import check_password_hash
+from email_validator import EmailNotValidError
+from sqlalchemy.exc import SQLAlchemyError, IntegrityError
+from configparser import NoOptionError
+from pbench.server.database.models.users import User
+from pbench.server.database.models.active_tokens import ActiveTokens
+
+from pbench.server.api.auth import Auth
+
+
+class RegisterUser(Resource):
+    """
+    Abstracted pbench API for registering a new user
+    """
+
+    def __init__(self, config, logger):
+        self.server_config = config
+        self.logger = logger
+        try:
+            self.token_expire_duration = self.server_config.get(
+                "pbench-server", "token_expiration_duration"
+            )
+        except NoOptionError:
+            self.token_expire_duration = None
+
+    def post(self):
+        """
+        Post request for registering a new user.
+        This requires a JSON data with required user fields
+        {
+            "username": "username",
+            "password": "password",
+            "first_name": first_name,
+            "last_name": "last_name",
+            "email": "user@domain.com"
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return: JSON Payload
+        if we succeed to add a user entry in database, the returned response_object will look like following:
+            response_object = {
+                    "status": "success", # not present if the request fails
+                    "message": "Successfully registered."/"failure message",
+                }
+        To get the auth token user has to perform the login action
+        """
+        # get the post data
+        user_data = request.get_json()
+        if not user_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        username = user_data.get("username")
+        if not username:
+            self.logger.warning("Missing username field")
+            abort(
+                400, message="Missing username field",
+            )
+        username = username.lower()
+        if username == "admin":
```

Comment:
Right now, I have kept this to avoid anyone simply using `admin` as a username and potentially avoid conflict in future when we have the real admin interface. We will  need to update the code when we add the role based authentication and we have a user with admin privileges.

Reply:
@ndokos, I believe that this function is for registering new users (the login happens elsewhere, once this request returns successfully).  So, the code is not allowing people to create an `admin` user, as this is a reserved account (although, how it actually gets created and maintained is TBD, I think).